File: api/src/modules/networking/networking.service.ts
import prisma from '../../lib/prismaClient';
import { AppError } from '../../middleware/errorHandler';
import { ConnectionStatus, SwipeDirection } from '@prisma/client';
import { Storage } from '@google-cloud/storage';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';

class NetworkingService {
  private storage: Storage;

  constructor() {
    this.storage = new Storage({
      projectId: process.env.GOOGLE_CLOUD_PROJECT_ID,
      keyFilename: process.env.GOOGLE_CLOUD_KEY_FILE,
    });
  }

  async updateProfile(userId: number, data: any) {
    const profile = await prisma.networkProfile.upsert({
      where: { userId },
      create: { userId, ...data },
      update: data,
    });
    return profile;
  }

  async createOrUpdateProfile(userId: number, profileData: any) {
    const {
      firstName,
      lastName,
      birthday,
      gender,
      country,
      city,
      instagramHandle,
      interests,
      isProfilePublic
    } = profileData;

    // Update user basic info
    const user = await prisma.user.update({
      where: { id: userId },
      data: {
        firstName,
        lastName,
        birthday: new Date(birthday),
        gender,
        country,
        city,
        instagramHandle,
        interests,
        isProfilePublic: isProfilePublic ?? true,
      },
    });

    // Create or update network profile
    const networkProfile = await prisma.networkProfile.upsert({
      where: { userId },
      create: {
        userId,
        visibility: isProfilePublic ? 'PUBLIC' : 'PRIVATE',
        lookingFor: 'FRIENDS',
        bio: '',
        interests,
      },
      update: {
        visibility: isProfilePublic ? 'PUBLIC' : 'PRIVATE',
        interests,
      },
    });

    return {
      user: {
        id: user.id,
        firstName: user.firstName,
        lastName: user.lastName,
        email: user.email,
        birthday: user.birthday,
        gender: user.gender,
        country: user.country,
        city: user.city,
        instagramHandle: user.instagramHandle,
        interests: user.interests,
        isProfilePublic: user.isProfilePublic,
        profilePhoto: user.profilePhoto,
      },
      networkProfile,
    };
  }

  async uploadProfilePhoto(userId: number, file: any) {
    try {
      const bucketName = process.env.GOOGLE_CLOUD_STORAGE_BUCKET || 'wizard-media';
      const fileName = `profile-photos/${userId}/${uuidv4()}${path.extname(file.originalname)}`;
      
      const bucket = this.storage.bucket(bucketName);
      const fileUpload = bucket.file(fileName);

      const stream = fileUpload.createWriteStream({
        metadata: {
          contentType: file.mimetype,
        },
        resumable: false,
      });

      return new Promise((resolve, reject) => {
        stream.on('error', (err) => {
          console.error('❌ [GCS] Upload error:', err);
          reject(new AppError('Failed to upload photo', 500));
        });

        stream.on('finish', async () => {
          try {
            // Make the file public
            await fileUpload.makePublic();
            
            const publicUrl = `https://storage.googleapis.com/${bucketName}/${fileName}`;
            
            // Update user's profile photo URL
            await prisma.user.update({
              where: { id: userId },
              data: { profilePhoto: publicUrl },
            });

            console.log('✅ [GCS] Photo uploaded successfully:', publicUrl);
            resolve({ photoUrl: publicUrl });
          } catch (err) {
            console.error('❌ [GCS] Error updating database:', err);
            reject(new AppError('Failed to save photo URL', 500));
          }
        });

        stream.end(file.buffer);
      });
    } catch (error) {
      console.error('❌ [GCS] Upload error:', error);
      throw new AppError('Failed to upload photo', 500);
    }
  }

  async getMyProfile(userId: number) {
    const user = await prisma.user.findUnique({
      where: { id: userId },
      select: {
        id: true,
        firstName: true,
        lastName: true,
        email: true,
        birthday: true,
        gender: true,
        country: true,
        city: true,
        instagramHandle: true,
        interests: true,
        isProfilePublic: true,
        profilePhoto: true,
        networkProfile: true,
      },
    });

    if (!user) {
      throw new AppError('User not found', 404);
    }

    return user;
  }

  async discover(userId: number, limit: number = 20) {
    const swipedUserIds = await prisma.userSwipe.findMany({
      where: { swiperId: userId },
      select: { swipedUserId: true },
    });
    const swipedIds = swipedUserIds.map(s => s.swipedUserId);

    const users = await prisma.user.findMany({
      where: {
        id: { notIn: [...swipedIds, userId] },
        isProfilePublic: true,
      },
      take: limit,
      select: {
        id: true,
        firstName: true,
        lastName: true,
        profilePhoto: true,
        bio: true,
        city: true,
        country: true,
        interests: true,
        networkProfile: true,
      },
    });

    return users;
  }

  async recordSwipe(userId: number, swipedUserId: number, direction: SwipeDirection) {
    await prisma.userSwipe.create({
      data: {
        swiperId: userId,
        swipedUserId,
        direction,
      },
    });

    // If right swipe, check for match
    if (direction === 'RIGHT') {
      const reciprocalSwipe = await prisma.userSwipe.findUnique({
        where: {
          swiperId_swipedUserId: {
            swiperId: swipedUserId,
            swipedUserId: userId,
          },
        },
      });

      if (reciprocalSwipe && reciprocalSwipe.direction === 'RIGHT') {
        // Create connection
        await prisma.connection.create({
          data: {
            requesterId: userId,
            receiverId: swipedUserId,
            status: 'ACCEPTED',
            acceptedAt: new Date(),
          },
        });
        return { matched: true };
      }
    }

    return { matched: false };
  }

  async getConnections(userId: number, status?: ConnectionStatus) {
    const where: any = {
      OR: [{ requesterId: userId }, { receiverId: userId }],
    };

    if (status) {
      where.status = status;
    }

    const connections = await prisma.connection.findMany({
      where,
      include: {
        requester: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            profilePhoto: true,
            city: true,
            country: true,
          },
        },
        receiver: {
          select: {
            id: true,
            firstName: true,
            lastName: true,
            profilePhoto: true,
            city: true,
            country: true,
          },
        },
      },
    });

    return connections.map(conn => ({
      ...conn,
      user: conn.requesterId === userId ? conn.receiver : conn.requester,
    }));
  }

  async sendConnectionRequest(userId: number, targetUserId: number) {
    const connection = await prisma.connection.create({
      data: {
        requesterId: userId,
        receiverId: targetUserId,
        status: 'PENDING',
      },
    });
    return connection;
  }

  async acceptConnection(userId: number, connectionId: number) {
    const connection = await prisma.connection.findUnique({
      where: { id: connectionId },
    });

    if (!connection || connection.receiverId !== userId) {
      throw new AppError('Connection not found', 404);
    }

    return prisma.connection.update({
      where: { id: connectionId },
      data: { status: 'ACCEPTED', acceptedAt: new Date() },
    });
  }

  async rejectConnection(userId: number, connectionId: number) {
    const connection = await prisma.connection.findUnique({
      where: { id: connectionId },
    });

    if (!connection || connection.receiverId !== userId) {
      throw new AppError('Connection not found', 404);
    }

    return prisma.connection.update({
      where: { id: connectionId },
      data: { status: 'REJECTED' },
    });
  }

  async searchUsers(query: string, limit: number = 20) {
    const users = await prisma.user.findMany({
      where: {
        isProfilePublic: true,
        OR: [
          { firstName: { contains: query, mode: 'insensitive' } },
          { lastName: { contains: query, mode: 'insensitive' } },
          { city: { contains: query, mode: 'insensitive' } },
        ],
      },
      take: limit,
      select: {
        id: true,
        firstName: true,
        lastName: true,
        profilePhoto: true,
        city: true,
        country: true,
        bio: true,
      },
    });

    return users;
  }

  async getUserProfile(userId: number, targetUserId: number) {
    const user = await prisma.user.findUnique({
      where: { id: targetUserId },
      select: {
        id: true,
        firstName: true,
        lastName: true,
        profilePhoto: true,
        bio: true,
        city: true,
        country: true,
        interests: true,
        instagramHandle: true,
        networkProfile: true,
      },
    });

    if (!user) {
      throw new AppError('User not found', 404);
    }

    // Record view
    await prisma.userView.create({
      data: {
        viewerId: userId,
        viewedUserId: targetUserId,
      },
    }).catch(() => {}); // Ignore if already viewed

    return user;
  }

  async updateLocation(userId: number, latitude: number, longitude: number) {
    await prisma.networkProfile.upsert({
      where: { userId },
      create: {
        userId,
        latitude,
        longitude,
        lastLocationUpdate: new Date(),
      },
      update: {
        latitude,
        longitude,
        lastLocationUpdate: new Date(),
      },
    });

    return { message: 'Location updated' };
  }
}

export default new NetworkingService();

File: mobile/lib/services/networking_service.dart
import 'dart:io';
import '../models/network_user.dart';
import '../models/connection.dart';
import 'api_client.dart';

class NetworkingService {
  final ApiClient _apiClient;

  NetworkingService(this._apiClient);

  /// Discover users for swiping
  Future<List<NetworkUser>> discoverUsers({int limit = 20}) async {
    try {
      final response = await _apiClient.get('/networking/discover', params: {'limit': limit.toString()});
      final List<dynamic> usersData = response['data'] ?? [];
      return usersData.map((userData) => NetworkUser.fromJson(userData)).toList();
    } catch (e) {
      print('❌ [NetworkingService] Error discovering users: $e');
      rethrow;
    }
  }

  /// Record a swipe action
  Future<Map<String, dynamic>> recordSwipe({
    required int swipedUserId,
    required String direction, // 'LEFT' or 'RIGHT'
  }) async {
    try {
      final response = await _apiClient.post('/networking/swipe', body: {
        'swipedUserId': swipedUserId,
        'direction': direction,
      });
      return response['data'] ?? {};
    } catch (e) {
      print('❌ [NetworkingService] Error recording swipe: $e');
      rethrow;
    }
  }

  /// Search users by query
  Future<List<NetworkUser>> searchUsers(String query) async {
    try {
      final response = await _apiClient.get('/networking/search', params: {'q': query});
      final List<dynamic> usersData = response['data'] ?? [];
      return usersData.map((userData) => NetworkUser.fromJson(userData)).toList();
    } catch (e) {
      print('❌ [NetworkingService] Error searching users: $e');
      rethrow;
    }
  }

  /// Get user profile by ID
  Future<NetworkUser> getUserProfile(int userId) async {
    try {
      final response = await _apiClient.get('/networking/user/$userId');
      return NetworkUser.fromJson(response['data']);
    } catch (e) {
      print('❌ [NetworkingService] Error getting user profile: $e');
      rethrow;
    }
  }

  /// Get all connections
  Future<List<Connection>> getConnections() async {
    try {
      final response = await _apiClient.get('/networking/connections');
      final List<dynamic> connectionsData = response['data'] ?? [];
      return connectionsData.map((connData) => Connection.fromJson(connData)).toList();
    } catch (e) {
      print('❌ [NetworkingService] Error getting connections: $e');
      rethrow;
    }
  }

  /// Send connection request
  Future<Connection> sendConnectionRequest(int targetUserId) async {
    try {
      final response = await _apiClient.post('/networking/connections/request', body: {
        'targetUserId': targetUserId,
      });
      return Connection.fromJson(response['data']);
    } catch (e) {
      print('❌ [NetworkingService] Error sending connection request: $e');
      rethrow;
    }
  }

  /// Accept connection request
  Future<Connection> acceptConnection(int connectionId) async {
    try {
      final response = await _apiClient.post('/networking/connections/$connectionId/accept');
      return Connection.fromJson(response['data']);
    } catch (e) {
      print('❌ [NetworkingService] Error accepting connection: $e');
      rethrow;
    }
  }

  /// Reject connection request
  Future<Connection> rejectConnection(int connectionId) async {
    try {
      final response = await _apiClient.post('/networking/connections/$connectionId/reject');
      return Connection.fromJson(response['data']);
    } catch (e) {
      print('❌ [NetworkingService] Error rejecting connection: $e');
      rethrow;
    }
  }

  /// Update network profile
  Future<Map<String, dynamic>> updateNetworkProfile({
    String? visibility,
    String? lookingFor,
    String? bio,
    List<String>? interests,
  }) async {
    try {
      final response = await _apiClient.put('/networking/profile', body: {
        if (visibility != null) 'visibility': visibility,
        if (lookingFor != null) 'lookingFor': lookingFor,
        if (bio != null) 'bio': bio,
        if (interests != null) 'interests': interests,
      });
      return response['data'] ?? {};
    } catch (e) {
      print('❌ [NetworkingService] Error updating network profile: $e');
      rethrow;
    }
  }

  /// Update user location
  Future<Map<String, dynamic>> updateLocation({
    required double latitude,
    required double longitude,
  }) async {
    try {
      final response = await _apiClient.post('/networking/location', body: {
        'latitude': latitude,
        'longitude': longitude,
      });
      return response['data'] ?? {};
    } catch (e) {
      print('❌ [NetworkingService] Error updating location: $e');
      rethrow;
    }
  }

  /// Create or update user profile
  Future<Map<String, dynamic>> createOrUpdateProfile(Map<String, dynamic> profileData) async {
    try {
      final response = await _apiClient.post('/networking/profile', body: profileData);
      return response['data'] ?? {};
    } catch (e) {
      print('❌ [NetworkingService] Error creating/updating profile: $e');
      rethrow;
    }
  }

  /// Upload profile photo
  Future<Map<String, dynamic>> uploadProfilePhoto(File imageFile) async {
    try {
      final response = await _apiClient.uploadFile('/networking/profile/photo', imageFile);
      return response['data'] ?? {};
    } catch (e) {
      print('❌ [NetworkingService] Error uploading profile photo: $e');
      rethrow;
    }
  }

  /// Get user's own profile
  Future<Map<String, dynamic>> getMyProfile() async {
    try {
      final response = await _apiClient.get('/networking/profile');
      return response['data'] ?? {};
    } catch (e) {
      print('❌ [NetworkingService] Error getting profile: $e');
      rethrow;
    }
  }
}
